internal/zinc-core/src/main/scala/sbt/internal/inc/IncrementalCommon.scala

```diff
@@ -501,13 +502,14 @@ private[inc] abstract class IncrementalCommon(
       Set.empty
     } else {
       if (invalidateTransitively) {
-        val firstClassTransitiveInvalidation = includeTransitiveInitialInvalidations(
-          initial,
-          IncrementalCommon.transitiveDeps(initial, log)(dependsOnClass),
-          dependsOnClass
-        )
+        // NOTE: As member reference relations do not include local relations, this invalidation will fully propagate
+        // thus we can't rely solely on `firstClassTransitiveInvalidation`. Better bet is to try to find transitive
+        // dependencies from result of `firstClassInvalidation`
+        val firstClassTransitiveInvalidation =
+          IncrementalCommon.transitiveDeps(firstClassInvalidation, log)(dependsOnClass)
         log.debug("Invalidate by brute force:\n\t" + firstClassTransitiveInvalidation)
-        firstClassTransitiveInvalidation ++ secondClassInvalidation ++ thirdClassInvalidation ++ recompiledClasses
+        firstClassInvalidation ++ firstClassTransitiveInvalidation ++ secondClassInvalidation ++
```

Comment:
Now that I think of it, it may actually be enough to have 
`firstClassTransitiveInvalidation ++ secondClassInvalidation ++ thirdClassInvalidation ++ recompiledClasses` as `firstClassTransitiveInvalidations` is now computed from `firstClassInvalidation`